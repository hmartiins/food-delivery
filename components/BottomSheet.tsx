import { ReactNode } from 'react';

import { Dimensions, Pressable } from 'react-native';

import { BlurView } from 'expo-blur';

import { MaterialIcons } from '@expo/vector-icons';

import { Gesture, GestureDetector } from 'react-native-gesture-handler';
import Animated, {
  FadeIn,
  FadeOut,
  SlideInDown,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
  withTiming,
} from 'react-native-reanimated';

import { Portal } from './Portal';

const { width } = Dimensions.get('window');
const sheetOverDrag = 20;

type Props = {
  height: number;
  children: ReactNode;
  onClose: () => void;
};

export const BottomSheet = ({ onClose, children, height }: Props) => {
  const offset = useSharedValue(0);

  function close() {
    offset.value = 0;
    onClose();
  }

  const gesture = Gesture.Pan()
    .onChange(event => {
      const offsetDelta = event.changeY + offset.value;
      const clamp = Math.max(-sheetOverDrag, offsetDelta);

      offset.value = offsetDelta > 0 ? offsetDelta : withSpring(clamp);
    })
    .onFinalize(() => {
      if (offset.value < height / 3) {
        offset.value = withSpring(0);
      } else {
        offset.value = withTiming(height, {}, () => runOnJS(close)());
      }
    });

  const translateY = useAnimatedStyle(() => ({
    transform: [{ translateY: offset.value }],
  }));

  return (
    <Portal name="bottom-sheet">
      <Animated.View
        testID="bottom-sheet-overlay"
        entering={FadeIn.duration(200)}
        exiting={FadeOut.duration(200)}
        className="absolute inset-0 z-40"
      >
        <BlurView
          testID="blur-view"
          intensity={20}
          tint="dark"
          className="flex-1"
        >
          <Pressable
            testID="bottom-sheet-backdrop"
            className="flex-1 bg-black/20"
            onPress={onClose}
          />
        </BlurView>
      </Animated.View>

      <GestureDetector gesture={gesture}>
        <Animated.View
          testID="bottom-sheet-container"
          entering={SlideInDown.springify()}
          className={`absolute left-0 right-0 z-50 rounded-t-3xl bg-white`}
          style={[
            {
              width,
              height,
              bottom: -sheetOverDrag * 1.3,
            },
            translateY,
          ]}
        >
          <MaterialIcons
            testID="bottom-sheet-handle"
            className="mt-0 self-center"
            name="horizontal-rule"
            color="gray"
            size={24}
          />
          {children}
        </Animated.View>
      </GestureDetector>
    </Portal>
  );
};
