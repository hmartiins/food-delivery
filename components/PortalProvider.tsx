import { Fragment, ReactNode } from 'react';

import { usePortalStore } from '@/stores/portal.store';

interface PortalProviderProps {
  children: ReactNode;
}

export const PortalProvider = ({ children }: PortalProviderProps) => {
  const { getComponents } = usePortalStore();
  const components = getComponents();

  return (
    <Fragment>
      {children}
      <Fragment>
        {components.map(([name, Component]) => (
          <Fragment key={name}>{Component}</Fragment>
        ))}
      </Fragment>
    </Fragment>
  );
};
