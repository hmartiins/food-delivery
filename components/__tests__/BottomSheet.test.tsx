import { Text, View } from 'react-native';

import { act, fireEvent, render, screen } from '@testing-library/react-native';

import { renderWithPortal } from '@/helpers';

import { BottomSheet } from '../BottomSheet';
import { PortalProvider } from '../PortalProvider';

jest.mock('react-native-reanimated', () => ({
  ...require('react-native-reanimated/mock'),
  useSharedValue: jest.fn(() => ({ value: 0 })),
  useAnimatedStyle: jest.fn(() => ({})),
  withSpring: jest.fn(),
  withTiming: jest.fn(),
  runOnJS: jest.fn(fn => fn),
}));

jest.mock('react-native-gesture-handler', () => {
  return {
    Gesture: {
      Pan: jest.fn(() => ({
        onChange: jest.fn().mockReturnThis(),
        onFinalize: jest.fn().mockReturnThis(),
      })),
    },
    GestureDetector: ({ children }: { children: React.ReactNode }) => children,
  };
});

jest.mock('@expo/vector-icons', () => ({
  MaterialIcons: (props: any) => 'MockedMaterialIcon',
}));

describe('<BottomSheet />', () => {
  const defaultProps = {
    height: 400,
    onClose: jest.fn(),
    children: <Text testID="test-content">Test Content</Text>,
  };

  beforeEach(() => jest.clearAllMocks());

  describe('Rendering', () => {
    it('should render without errors', () => {
      renderWithPortal(<BottomSheet {...defaultProps} />);

      expect(screen.getByTestId('bottom-sheet-overlay')).toBeTruthy();
      expect(screen.getByTestId('bottom-sheet-container')).toBeTruthy();
      expect(screen.getByTestId('bottom-sheet-backdrop')).toBeTruthy();
    });

    it('should render children correctly', () => {
      renderWithPortal(<BottomSheet {...defaultProps} />);

      expect(screen.getByTestId('test-content')).toBeTruthy();
      expect(screen.getByText('Test Content')).toBeTruthy();
    });

    it('should render with different children', () => {
      const customChildren = (
        <View testID="custom-content">
          <Text>Custom Title</Text>
          <Text>Custom Subtitle</Text>
        </View>
      );

      renderWithPortal(
        <BottomSheet {...defaultProps}>{customChildren}</BottomSheet>
      );

      expect(screen.getByTestId('custom-content')).toBeTruthy();
      expect(screen.getByText('Custom Title')).toBeTruthy();
      expect(screen.getByText('Custom Subtitle')).toBeTruthy();
    });
  });

  describe('Props', () => {
    it('should use correct height prop', () => {
      const customHeight = 600;
      renderWithPortal(<BottomSheet {...defaultProps} height={customHeight} />);

      const container = screen.getByTestId('bottom-sheet-container');
      expect(container.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            height: customHeight,
          }),
        ])
      );
    });

    it('should handle different height values', () => {
      const heights = [200, 400, 600, 800];

      heights.forEach(height => {
        const { unmount } = renderWithPortal(
          <BottomSheet {...defaultProps} height={height} />
        );

        const container = screen.getByTestId('bottom-sheet-container');
        expect(container.props.style).toEqual(
          expect.arrayContaining([
            expect.objectContaining({
              height,
            }),
          ])
        );

        unmount();
      });
    });
  });

  describe('Interactions', () => {
    it('should call onClose when backdrop is pressed', async () => {
      const mockOnClose = jest.fn();
      renderWithPortal(<BottomSheet {...defaultProps} onClose={mockOnClose} />);

      const backdrop = screen.getByTestId('bottom-sheet-backdrop');
      await act(async () => fireEvent.press(backdrop));

      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it('should not call onClose when container is pressed', async () => {
      const mockOnClose = jest.fn();
      renderWithPortal(<BottomSheet {...defaultProps} onClose={mockOnClose} />);

      const container = screen.getByTestId('bottom-sheet-container');
      await act(async () => fireEvent.press(container));

      expect(mockOnClose).not.toHaveBeenCalled();
    });

    it('should not call onClose when children are pressed', async () => {
      const mockOnClose = jest.fn();
      renderWithPortal(<BottomSheet {...defaultProps} onClose={mockOnClose} />);

      const content = screen.getByTestId('test-content');
      await act(async () => fireEvent.press(content));

      expect(mockOnClose).not.toHaveBeenCalled();
    });
  });
});

describe('<BottomSheet /> Snapshots', () => {
  const defaultProps = {
    height: 400,
    onClose: jest.fn(),
    children: <Text testID="test-content">Test Content</Text>,
  };

  it('should match snapshot with default props', () => {
    const component = render(
      <PortalProvider>
        <BottomSheet {...defaultProps} />
      </PortalProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with different height', () => {
    const component = render(
      <PortalProvider>
        <BottomSheet {...defaultProps} height={600} />
      </PortalProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot with complex children', () => {
    const complexChildren = (
      <View testID="complex-content">
        <Text>Title</Text>
        <Text>Subtitle</Text>
        <View testID="nested-view">
          <Text>Nested content</Text>
        </View>
      </View>
    );

    const component = render(
      <PortalProvider>
        <BottomSheet {...defaultProps}>{complexChildren}</BottomSheet>
      </PortalProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
