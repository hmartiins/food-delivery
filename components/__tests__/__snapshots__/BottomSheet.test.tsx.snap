// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BottomSheet /> Snapshots should match snapshot with complex children 1`] = `
[
  <View
    className="absolute inset-0 z-40"
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    testID="bottom-sheet-overlay"
  >
    <View
      className="flex-1"
      style={
        [
          {
            "backgroundColor": "transparent",
          },
          undefined,
        ]
      }
      testID="blur-view"
    >
      <ViewManagerAdapter_ExpoBlurView
        blurReductionFactor={4}
        experimentalBlurMethod="none"
        intensity={20}
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        tint="dark"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="flex-1 bg-black/20"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="bottom-sheet-backdrop"
      />
    </View>
  </View>,
  <View
    className="absolute left-0 right-0 z-50 rounded-t-3xl bg-white"
    entering={BaseAnimationMock {}}
    style={
      [
        {
          "bottom": -26,
          "height": 400,
          "width": 750,
        },
        {},
      ]
    }
    testID="bottom-sheet-container"
  >
    <Text
      allowFontScaling={false}
      className="mt-0 self-center"
      selectable={false}
      style={
        [
          {
            "color": "gray",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "material",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="bottom-sheet-handle"
    >
      
    </Text>
    <View
      testID="complex-content"
    >
      <Text>
        Title
      </Text>
      <Text>
        Subtitle
      </Text>
      <View
        testID="nested-view"
      >
        <Text>
          Nested content
        </Text>
      </View>
    </View>
  </View>,
]
`;

exports[`<BottomSheet /> Snapshots should match snapshot with default props 1`] = `
[
  <View
    className="absolute inset-0 z-40"
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    testID="bottom-sheet-overlay"
  >
    <View
      className="flex-1"
      style={
        [
          {
            "backgroundColor": "transparent",
          },
          undefined,
        ]
      }
      testID="blur-view"
    >
      <ViewManagerAdapter_ExpoBlurView
        blurReductionFactor={4}
        experimentalBlurMethod="none"
        intensity={20}
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        tint="dark"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="flex-1 bg-black/20"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="bottom-sheet-backdrop"
      />
    </View>
  </View>,
  <View
    className="absolute left-0 right-0 z-50 rounded-t-3xl bg-white"
    entering={BaseAnimationMock {}}
    style={
      [
        {
          "bottom": -26,
          "height": 400,
          "width": 750,
        },
        {},
      ]
    }
    testID="bottom-sheet-container"
  >
    <Text
      allowFontScaling={false}
      className="mt-0 self-center"
      selectable={false}
      style={
        [
          {
            "color": "gray",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "material",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="bottom-sheet-handle"
    >
      
    </Text>
    <Text
      testID="test-content"
    >
      Test Content
    </Text>
  </View>,
]
`;

exports[`<BottomSheet /> Snapshots should match snapshot with different height 1`] = `
[
  <View
    className="absolute inset-0 z-40"
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    testID="bottom-sheet-overlay"
  >
    <View
      className="flex-1"
      style={
        [
          {
            "backgroundColor": "transparent",
          },
          undefined,
        ]
      }
      testID="blur-view"
    >
      <ViewManagerAdapter_ExpoBlurView
        blurReductionFactor={4}
        experimentalBlurMethod="none"
        intensity={20}
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        tint="dark"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        className="flex-1 bg-black/20"
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="bottom-sheet-backdrop"
      />
    </View>
  </View>,
  <View
    className="absolute left-0 right-0 z-50 rounded-t-3xl bg-white"
    entering={BaseAnimationMock {}}
    style={
      [
        {
          "bottom": -26,
          "height": 600,
          "width": 750,
        },
        {},
      ]
    }
    testID="bottom-sheet-container"
  >
    <Text
      allowFontScaling={false}
      className="mt-0 self-center"
      selectable={false}
      style={
        [
          {
            "color": "gray",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "material",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="bottom-sheet-handle"
    >
      
    </Text>
    <Text
      testID="test-content"
    >
      Test Content
    </Text>
  </View>,
]
`;
